xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdsbalancer
+
+import (
+	"errors"
+	"sync"
+
+	"google.golang.org/grpc/internal/buffer"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+)
+
+
+// clusterHandler will be given a name representing a cluster. It will then update the CDS policy
+// constantly with a list of Clusters to pass down to XdsClusterResolverLoadBalancingPolicyConfig
+// in a stream like fashion.
+type clusterHandler struct {
+	// A mutex to protect entire tree of clusters.
+	clusterMutex sync.Mutex
+	root         *clusterNode
+
+	// A way to ping CDS Balanccer about any updates or errors to a Node in the tree.
+	// This will either get called from this handler constructing an update or from a child with an error.
+	updateChannel *buffer.Unbounded
+
+	xdsClient      xdsClientInterface
+}
+
+func (ch *clusterHandler) updateRootCluster(clusterName string) /*Return value - the expectation for a stream back to the CDS Policy*/ {
+	if ch.root == nil {
+		// Construct a root node on first update.
+		ch.root = createClusterNode(clusterName, ch.xdsClient, ch)
+		return
+	}
+	// Check if root cluster was changed. If it was, delete old one and start new one, if not
+	// do nothing.
+	if clusterName != ch.root.clusterUpdate.ServiceName {
+		ch.root.delete()
+		ch.root = createClusterNode(clusterName, ch.xdsClient, ch)
+	}
+}
+
+
+// This function tries to construct a cluster update to send to CDS.
+func (ch *clusterHandler) constructClusterUpdate() {
+	ch.clusterMutex.Lock()
```

Comment:
This is going to cause deadlock, when `constructClusterUpdate` is called with `clusterMutex` held.

A simpler way to think of this is, only the "entrance" functions need to take the lock (in you code, `updateRootCluster` and `handleResp`).

Reply:
A side note, if you run into a deadlock while testing, `go test` takes flag `-timeout 3s`, the test will timeout after 3 seconds, and will print the stack trace. From that you should be able to see where the program is blocked on.

Reply:
In that case, I should go ahead and delete the mutex lock from delete(), right?

Reply:
I also got rid of delete() and deleteRecursive(), since the point of those was to grab the lock.